### PR TITLE
Call toFirestore() only once

### LINKED ISSUE #3742: Firestore: CollectionReference.withConverter(converter).add(data) invokes converter.toFirestore twice
### [REQUIRED] Describe your environment

  * Operating System version: macOS Catalina 10.15.6
  * Browser version: N/A
  * Firebase SDK version: 7.19.1
  * Firebase Product: firestore

### [REQUIRED] Describe the problem
`firebase.firestore().collection('...').withConverter(converter).add(data)` invokes `converter.toFirestore` twice while  `firebase.firestore().collection('...').withConverter(converter).doc().set(data)` invokes `converter.toFirestore` only once.
Is this intended behavior?

_Although this is not a desirable style of processing, if the toFirestore() process has a side effect on the input value, the second invocation fails (See example below)._

#### Steps to reproduce:
try to run following code please.

#### Relevant Code:
```
const firebase = require('firebase')

// Please use your firebase config here
firebase.initializeApp(firebaseConfig)

const document = {
  someArray: [
    { text: '' },
  ],
}

const converter = {
  toFirestore: (modelObject) => {
    console.log('toFirestore', modelObject)
    modelObject.someArray = 'text'
    return { text: modelObject.someArray[0].text }
  },
  fromFirestore: () => {
    throw new ReferenceError()
  },
}

// log 'toFirestore' twice, and then fail second conversion.
firebase.firestore().collection('posts')
  .withConverter(converter)
  .add(document)

// log 'toFirestore' only once and complete as expected.
firebase.firestore().collection('posts')
  .withConverter(converter)
  .doc()
  .set(document)
```

### PR DESCRIPTION
**Pull request by thebrianchen**
Fixes #3742.

### CODE REVIEW

File: packages/firestore/src/api/database.ts [hunk 1]
```diff
@@ -2375,12 +2381,10 @@ export class CollectionReference<T = DocumentData>
...
     validateExactNumberOfArgs('CollectionReference.add', arguments, 1);
-    const convertedValue = this._converter
-      ? this._converter.toFirestore(value)
-      : value;
-    validateArgType('CollectionReference.add', 'object', 1, convertedValue);
     const docRef = this.doc();
-    return docRef.set(value).then(() => docRef);
+    return ((docRef as unknown) as DocumentReference)
+      .set(value, undefined, 'CollectionReference.add')
+      .then(() => docRef);
```
Comment (schmidt-sebastian) [inline packages/firestore/src/api/database.ts:2387]: Couldn't you just leave the original code as is and do:

return docRef.withConverter(defaultConverter).set(value).then(() => docRef)

?

File: packages/firestore/src/api/database.ts [hunk 2]
```diff
@@ -2379,8 +2379,24 @@ export class CollectionReference<T = DocumentData>
...
+        return modelObject;
+      },
+      fromFirestore(
+        snapshot: QueryDocumentSnapshot,
+        options: SnapshotOptions
+      ): DocumentData {
+        return snapshot.data(options)!;
+      }
+    };
+    return docRef
```
Comment (schmidt-sebastian) [inline packages/firestore/src/api/database.ts:2396]: Can you just call the DocumentReference constructor here? That way you can pass "null" and don't need to specify a converter.

Comment (thebrianchen) [reply packages/firestore/src/api/database.ts]: Done.

File: packages/firestore/src/api/database.ts [hunk 3]
```diff
@@ -2379,8 +2379,18 @@ export class CollectionReference<T = DocumentData>
...
     const docRef = this.doc();
-    return docRef.set(value).then(() => docRef);
+
+    // Call set() with the converted value directly to avoid calling toFirestore() a second time.
+    // Cast to unknown in order to access the _key property.
+    return new DocumentReference(
+      ((docRef as unknown) as DocumentReference)._key,
+      this.firestore,
+      null
+    )
```
Comment (schmidt-sebastian) [inline packages/firestore/src/api/database.ts:2391]:
```suggestion
    return new DocumentReference(
      (docRef as DocumentReference<T>)._key,
      this.firestore,
      null
    )
```

**Review by schmidt-sebastian — APPROVED** [review]

**Pull request merged**